Fix URL parsing.

ADmad opened the pull request:
Recheck whether a route is static when parsing a URL. Extension parsing can be enabled after a route was added to the collection and it can no longer be treated as a static route.

Refs #17081

<!---

Please describe the reason for the changes you're proposing. If it is a large change, please open an issue to discuss first.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Fix URL parsing.

Recheck whether a route is static when parsing a URL.
Extension parsing can be enabled after a route was added to
the collection and it can no longer be treated as a static route.

ndm2 commented on src/Routing/RouteCollection.php:
```diff
@@ -147,6 +147,12 @@ public function parse(string $url, string $method = ''): array
             foreach ($this->staticPaths[$decoded] as $route) {
+                // Extension parsing was enabled after the route was added to
+                // the collection and it can no longer be treated as a static route.
+                if ($route->getExtensions()) {
```
Looks like this will never be reached, as `/foo.json` wouldn't be in the static paths array, but only `/foo`?

ADmad replied:
Yes, this is not necessary, I will remove it. It actually prevents early parsing of `/foo` URL which would also match the same route.

ndm2 commented:
I guess we can't rely on extensions being in `.text` format, as parsing could be overwritten? Meaning splitting of parts of the request path for lookup isn't an option?

markstory committed:
Expand test coverage and simplify code.

markstory commented:
Looks good to me. I expanded the tests and simplified the code a bit.

ndm2 commented on src/Routing/RouteCollection.php:
```diff
@@ -147,6 +147,12 @@ public function parse(string $url, string $method = ''): array
 
         if (isset($this->staticPaths[$decoded])) {
```
I just noticed, shouldn't this rather use a string with the possible query parameters removed?

ADmad replied:
`$decoded` is the string after possible query params removed, that bit of code is not visible in the diff.

ndm2 replied:
The query parameters aren't removed from `$decoded` AFAICT tell, but from `$url`. I think the decoding part may need to be moved below the query parameters removal part.

ADmad replied:
My bad, yes the query params are not removed from `$decoded`. I'll look into it.

That said we need to deprecate `RouterCollection::parse()`, it's actually unused. `RoutingMiddleware` using `RouteCollection::parseRequest()`.

ndm2 replied:
Yeah, should've already been deprecated back in 3.x I guess. A test for both `parse()` and `parseRequest()` using a static path with query string arguments would be good though, it seems that the existing ones using query strings are used with dynamic paths only.

markstory replied:
> Yeah, should've already been deprecated back in 3.x I guess. 

Yep I missed this back then. Lets get rid of it in 5.x. I'll take care of this.

ADmad committed:
Optimize code.

Ignore extensions when checking if route is static since the route
can parse URLs with and without extensions.

ADmad commented:
I also removed the extensions check when deciding whether a route is static since a route connected as `$builder->connect('/foo')->setExtensions(['json'])` can parse both `/foo` and `/foo.json`. So having it in the static array speeds up parsing of `/foo`.

ADmad committed:
Remove query params before checking for static patch match.
